bluesky_queueserver/manager/worker.py

```diff
@@ -222,126 +228,196 @@ def plan():
+        (no messages can be sent to the closed environment).
+        """
+        err_msg = ""
+        if self._exit_event.is_set():
+            self._exit_confirmed_event.set()
+            status = "accepted"
+        else:
+            status = "rejected"
+            err_msg = "Environment closing was not initiated. Use command 'command_close_env' " \
+                      "to initiate closing and wait for RE Worker state: " \
+                      "self._state['environment_state']=='closing'"
+        msg_out = {"status": status, "err_msg": err_msg}
+        return msg_out
+
+    def _command_run_plan_handler(self, *, plan_info):
+        """
+        Initiate execution of a new plan.
+        """
+        logger.info("Starting execution of a plan")
+        # TODO: refine the criteria of acceptance of the new plan.
+        invalid_state = 0
+        if not self._execution_queue.empty():
+            invalid_state = 1
+        elif self._RE._state == 'running':
+            invalid_state = 2
+        elif self._state["running_plan"] or self._state["running_plan_completed"]:
+            invalid_state = 3
+
+        err_msg = ""
+        if not invalid_state:  # == 0
+            try:
+                # Value is a dictionary with plan parameters
+                self._load_new_plan(plan_info=plan_info)
+                status = "accepted"
+            except Exception as ex:
+                status = "error"
+                err_msg = str(ex)
+        else:
+            status = "rejected"
+            msg_list = ["the execution queue is not empty",
+                        "another plan is running",
+                        "worker is not reset after completion of the previous plan"]
+            try:
+                s = msg_list[invalid_state - 1]
+            except Exception:
+                s = "UNDETERMINED CONDITION IS PRESENT"  # Shouldn't ever be printed
+            err_msg = \
+                f"Trying to run a plan (start Run Engine) while {s}.\n" \
+                "This may indicate a serious issue with the plan queue execution mechanism.\n" \
+                "Please report the issue to developers."
+
+        msg_out = {"status": status, "err_msg": err_msg}
+        return msg_out
+
+    def _command_pause_plan_handler(self, *, option):
+        """
+        Pause running plan. Options: `deferred` and `immediate`.
+        """
+        # Stop the loop in main thread
+        logger.info("Pausing Run Engine")
+        pausing_options = ("deferred", "immediate")
+        # TODO: the question is whether it is possible or should be allowed to pause a plan in
+        #       any other state than 'running'???
+        err_msg = ""
+        if self._RE._state == 'running':
+            try:
+                if option not in pausing_options:
+                    raise RuntimeError(f"Option '{option}' is not supported. "
+                                       f"Available options: {pausing_options}")
```

Comment:
Should this check be done outside of this condition?

Reply:
I don't think so. First the command should be checked at the manager, so this check should pass if there are no bugs. And RE is also expected to be running to be running when the command is issued if there are no bugs in the manager. It's just another layer of safety that protects RE from being driven into strange condition. In terms of performance, checking if RE is running seems less costly, so we do it first.